File: LeetCode/c++/ReverseLinkedList.cpp
#include <iostream>

using namespace std;

struct ListNode 
{
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

ListNode* reverseList(ListNode* head)
{
	if (head == NULL)
		return NULL;
	else
	{
		ListNode *node = head;
		ListNode *reversedNode = new ListNode(node->val);
		node = node->next;

		while (node != NULL)
		{
			ListNode *temp = new ListNode(node->val);
			temp->next = reversedNode;
			reversedNode = temp;

			node = node->next;
		}

		return reversedNode;
	}
}

int main()
{
	ListNode *head = new ListNode(7);
	ListNode *node = head;

	for (int i = 6; i>=0; i--)
	{
		ListNode *nextNode = new ListNode(i);
		node->next = nextNode;
		node = node->next;
	}

	node = reverseList(head);

	while (node != NULL)
	{
		cout << node->val << " ";
		node = node->next;
	}

	return 0;
}

File: LeetCode/iter1/python/ReverseLinkedList.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def reverseList(self, head):
        if(head == None or head.next == None):
            return head
        else:
            nextNode = head.next
            head.next = None #head的下一个节点设为none
            while(nextNode != None):        
                tempNode = nextNode.next
                nextNode.next = head
                head = nextNode
                nextNode = tempNode
                
            return head

#测试函数
s = Solution()
l1 = ListNode(1)
l2 = ListNode(2)
l3 = ListNode(3)
# l4 = ListNode(4)
# l5 = ListNode(5)

l1.next = l2
l2.next = l3
# l3.next = l4
# l4.next = l5

l = s.reverseList(l1)

while(l!=None):
    print(l.val)
    l = l.next